File: packages/shared-types/src/query.ts
// Query Contract Types
// Strict schema for structured ocean-data queries

export type SupportedLanguage = 'en-IN' | 'ml-IN' | 'hi-IN' | 'mr-IN';

export type Intent =
  | 'profile_search'
  | 'timeseries_summary'
  | 'depth_profile_summary'
  | 'anomaly_detection'
  | 'scenario_projection'
  | 'marine_condition_briefing'
  | 'dataset_explanation'
  | 'export_results';

export type RegionType = 'bbox' | 'radius' | 'polygon' | 'named_region' | 'route';

export interface BBoxRegion {
  type: 'bbox';
  min_lat: number;
  max_lat: number;
  min_lon: number;
  max_lon: number;
}

export interface RadiusRegion {
  type: 'radius';
  lat: number;
  lon: number;
  radius_km: number;
}

export interface PolygonRegion {
  type: 'polygon';
  coordinates: number[][][]; // GeoJSON polygon
}

export interface NamedRegion {
  type: 'named_region';
  name: 'arabian_sea' | 'bay_of_bengal' | 'kerala_coast' | 'indian_ocean' | 'equatorial_indian_ocean';
}

export interface RouteRegion {
  type: 'route';
  origin: { lat: number; lon: number };
  destination: { lat: number; lon: number };
  corridor_km?: number;
}

export type Region = BBoxRegion | RadiusRegion | PolygonRegion | NamedRegion | RouteRegion;

export interface TimeRange {
  start: string; // ISO 8601 date
  end: string;   // ISO 8601 date
}

export interface DepthRange {
  min: number; // meters
  max: number; // meters
}

export type Variable = 'temperature' | 'salinity' | 'oxygen' | 'chlorophyll' | 'nitrate' | 'ph';

export type QualityFilter = 'all' | 'recommended' | 'good_only';

export type Aggregation = 'profile' | 'daily' | 'weekly' | 'monthly' | 'depth_bin';

export interface StructuredQuery {
  intent: Intent;
  language: SupportedLanguage;
  region?: Region;
  time_range?: TimeRange;
  depth_range_m?: DepthRange;
  variables?: Variable[];
  quality_filter?: QualityFilter;
  aggregation?: Aggregation;
  limit?: number;
  
  // For marine_condition_briefing
  distance_km?: number;
  origin?: { lat: number; lon: number };
  destination?: { lat: number; lon: number };
  departure_time?: string; // ISO 8601
  vessel_type?: string;
  include_forecast?: boolean;
  
  // For anomaly_detection
  reference_period?: TimeRange;
  analysis_period?: TimeRange;
  depth_m?: number;
  threshold_std?: number;
  
  // For scenario_projection
  trend_window?: TimeRange;
  projection_years?: number;
  model?: 'linear_trend' | 'polynomial' | 'theil_sen';
  assumptions?: string[];
  
  // For export_results
  export_format?: 'profiles' | 'observations' | 'summary';
}

export interface QueryPlanResult {
  status: 'ready' | 'needs_clarification' | 'unsupported';
  intent: Intent;
  language: SupportedLanguage;
  query: StructuredQuery;
  clarification_question?: string;
  warnings: string[];
}

// Supported named regions with bounding boxes
export const NAMED_REGIONS: Record<NamedRegion['name'], BBoxRegion> = {
  arabian_sea: { type: 'bbox', min_lat: 8.0, max_lat: 25.0, min_lon: 60.0, max_lon: 78.0 },
  bay_of_bengal: { type: 'bbox', min_lat: 5.0, max_lat: 22.0, min_lon: 80.0, max_lon: 100.0 },
  kerala_coast: { type: 'bbox', min_lat: 8.0, max_lat: 13.0, min_lon: 74.0, max_lon: 77.0 },
  indian_ocean: { type: 'bbox', min_lat: -30.0, max_lat: 30.0, min_lon: 30.0, max_lon: 120.0 },
  equatorial_indian_ocean: { type: 'bbox', min_lat: -10.0, max_lat: 10.0, min_lon: 40.0, max_lon: 110.0 },
};

File: apps/api/app/schemas/query.py
# API Schemas - Query
from pydantic import BaseModel, Field
from typing import Optional, List, Literal, Union, Annotated
from datetime import datetime
from enum import Enum


class SupportedLanguage(str, Enum):
    EN_IN = "en-IN"
    HI_IN = "hi-IN"
    ML_IN = "ml-IN"
    TA_IN = "ta-IN"
    TE_IN = "te-IN"
    BN_IN = "bn-IN"
    GU_IN = "gu-IN"
    MR_IN = "mr-IN"
    OR_IN = "or-IN"
    KN_IN = "kn-IN"


class Intent(str, Enum):
    PROFILE_SEARCH = "profile_search"
    TIMESERIES_SUMMARY = "timeseries_summary"
    DEPTH_PROFILE_SUMMARY = "depth_profile_summary"
    ANOMALY_DETECTION = "anomaly_detection"
    SCENARIO_PROJECTION = "scenario_projection"
    MARINE_CONDITION_BRIEFING = "marine_condition_briefing"
    DATASET_EXPLANATION = "dataset_explanation"
    EXPORT_RESULTS = "export_results"


class RegionType(str, Enum):
    BBOX = "bbox"
    RADIUS = "radius"
    POLYGON = "polygon"
    NAMED_REGION = "named_region"
    ROUTE = "route"


class BBoxRegion(BaseModel):
    type: Literal[RegionType.BBOX] = RegionType.BBOX
    min_lat: float = Field(ge=-90, le=90)
    max_lat: float = Field(ge=-90, le=90)
    min_lon: float = Field(ge=-180, le=180)
    max_lon: float = Field(ge=-180, le=180)


class RadiusRegion(BaseModel):
    type: Literal[RegionType.RADIUS] = RegionType.RADIUS
    lat: float = Field(ge=-90, le=90)
    lon: float = Field(ge=-180, le=180)
    radius_km: float = Field(gt=0, le=5000)


class PolygonRegion(BaseModel):
    type: Literal[RegionType.POLYGON] = RegionType.POLYGON
    coordinates: List[List[List[float]]]  # GeoJSON polygon


class NamedRegion(BaseModel):
    type: Literal[RegionType.NAMED_REGION] = RegionType.NAMED_REGION
    name: Literal["arabian_sea", "bay_of_bengal", "kerala_coast", "indian_ocean", "equatorial_indian_ocean"]


class RouteRegion(BaseModel):
    type: Literal[RegionType.ROUTE] = RegionType.ROUTE
    origin: "Coordinate"
    destination: "Coordinate"
    corridor_km: Optional[float] = Field(default=None, gt=0)


class Coordinate(BaseModel):
    lat: float = Field(ge=-90, le=90)
    lon: float = Field(ge=-180, le=180)


Region = Annotated[
    Union[BBoxRegion, RadiusRegion, PolygonRegion, NamedRegion, RouteRegion],
    Field(discriminator="type"),
]


class TimeRange(BaseModel):
    start: str  # ISO 8601
    end: str    # ISO 8601


class DepthRange(BaseModel):
    min: float = Field(ge=0, le=6000)
    max: float = Field(ge=0, le=6000)


class Variable(str, Enum):
    TEMPERATURE = "temperature"
    SALINITY = "salinity"
    OXYGEN = "oxygen"
    CHLOROPHYLL = "chlorophyll"
    NITRATE = "nitrate"
    PH = "ph"


class QualityFilter(str, Enum):
    ALL = "all"
    RECOMMENDED = "recommended"
    GOOD_ONLY = "good_only"


class Aggregation(str, Enum):
    PROFILE = "profile"
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    DEPTH_BIN = "depth_bin"


class ModelType(str, Enum):
    LINEAR_TREND = "linear_trend"
    POLYNOMIAL = "polynomial"
    THEIL_SEN = "theil_sen"


class ExportFormat(str, Enum):
    PROFILES = "profiles"
    OBSERVATIONS = "observations"
    SUMMARY = "summary"


class StructuredQuery(BaseModel):
    intent: Intent
    language: SupportedLanguage = SupportedLanguage.EN_IN
    region: Optional[Region] = None
    time_range: Optional[TimeRange] = None
    depth_range_m: Optional[DepthRange] = None
    variables: Optional[List[Variable]] = None
    quality_filter: QualityFilter = QualityFilter.RECOMMENDED
    aggregation: Aggregation = Aggregation.PROFILE
    limit: int = Field(default=500, ge=1, le=5000)

    # Marine condition briefing
    distance_km: Optional[float] = Field(default=None, gt=0)
    origin: Optional[Coordinate] = None
    destination: Optional[Coordinate] = None
    departure_time: Optional[str] = None
    vessel_type: Optional[str] = None
    include_forecast: bool = True

    # Anomaly detection
    reference_period: Optional[TimeRange] = None
    analysis_period: Optional[TimeRange] = None
    depth_m: Optional[float] = Field(default=None, ge=0, le=6000)
    threshold_std: float = Field(default=2.0, gt=0)

    # Scenario projection
    trend_window: Optional[TimeRange] = None
    projection_years: Optional[int] = Field(default=None, ge=1, le=50)
    model: ModelType = ModelType.LINEAR_TREND
    assumptions: List[str] = Field(default_factory=list)

    # Export
    export_format: ExportFormat = ExportFormat.PROFILES


class QueryPlanRequest(BaseModel):
    message: str
    language: Optional[SupportedLanguage] = None


class QueryPlanResponse(BaseModel):
    status: Literal["ready", "needs_clarification", "unsupported"]
    intent: Intent
    language: SupportedLanguage
    query: StructuredQuery
    clarification_question: Optional[str] = None
    warnings: List[str] = Field(default_factory=list)